### PR TITLE
DBZ-6366: Code Improvements for skip.messages.without.change

### PR DESCRIPTION
**Pull request by ironakj**
Addresses/Closes DBZ-6366

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -460,7 +460,7 @@ private void createDataEventsForDataCollection(MongoDbPartition partition) throw
             Object[] firstRow = null;
 
             for (BsonDocument doc : collection.find(predicate).sort(new Document(DOCUMENT_ID, 1))
-                    .limit(connectorConfig.getIncrementalSnashotChunkSize())) {
+                    .limit(connectorConfig.getIncrementalSnapshotChunkSize())) {
```
Comment (ironakj) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotChangeEventSource.java:463]: Typo in method name fix

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -176,7 +176,7 @@ protected void sendEvent(P partition, EventDispatcher<P, T> dispatcher, OffsetCo
      */
     protected ChangeRecordEmitter<P> getChangeRecordEmitter(P partition, T dataCollectionId,
                                                             OffsetContext offsetContext, Object[] row) {
-        return new SnapshotChangeRecordEmitter<>(partition, offsetContext, row, clock);
+        return new SnapshotChangeRecordEmitter<>(partition, offsetContext, row, clock, connectorConfig);
```
Comment (ironakj) [inline debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:179]: Not really necessary here, should I create a different constructor instead? @Naros @jpechane

File: debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java [hunk 3]
```diff
@@ -24,10 +25,13 @@
     private final OffsetContext offsetContext;
     private final Clock clock;
 
-    public AbstractChangeRecordEmitter(P partition, OffsetContext offsetContext, Clock clock) {
+    private final CommonConnectorConfig connectorConfig;
+
+    public AbstractChangeRecordEmitter(P partition, OffsetContext offsetContext, Clock clock, CommonConnectorConfig connectorConfig) {
```
Comment (Naros) [inline debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java:30]: I'm not a fan of pushing `CommonConnectorConfig` and the `skipMessagesWithoutChange` method up this far into the class hierarchy as it forces MongoDB to explicitly return false, making most of these changes at this level in the class hierarchy a bit superfluous.

I'd prefer you only pass `CommonConnectorConfig` up to the `RelationalChangeRecordEmitter` and simply expose the `skipMessagesWithoutChange` there, which gives a clear indication that this is a relational connector-only feature as it isn't exposed at the top of the class hierarchy that both relational and non-relational connectors are based from.

**Review by Naros — CHANGES_REQUESTED** [review]
One comment inline.

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java (hunk 2 above)
Comment (Naros) [reply debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:179]: I think it's OK to add the constructor argument that way, but I made a suggestion below regarding how far up the class hierarchy this should be passed.

File: debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java (hunk 3 above)
Comment (Naros) [reply debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java:30]: After reviewing the Jira issue, it seems we may plan to extend this to MongoDB, and in that case, this makes sense.

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java [hunk 4]
```diff
@@ -33,8 +34,8 @@
     public static final String PK_UPDATE_OLDKEY_FIELD = "__debezium.oldkey";
     public static final String PK_UPDATE_NEWKEY_FIELD = "__debezium.newkey";
 
-    public RelationalChangeRecordEmitter(P partition, OffsetContext offsetContext, Clock clock) {
-        super(partition, offsetContext, clock);
+    public RelationalChangeRecordEmitter(P partition, OffsetContext offsetContext, Clock clock, CommonConnectorConfig connectorConfig) {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java:37]: @ironakj Can we use `RelationalDatabaseConnectorConfig` here?

File: debezium-core/src/main/java/io/debezium/relational/SnapshotChangeRecordEmitter.java [hunk 5]
```diff
@@ -19,8 +20,8 @@
 
     private final Object[] row;
 
-    public SnapshotChangeRecordEmitter(P partition, OffsetContext offset, Object[] row, Clock clock) {
-        super(partition, offset, clock);
+    public SnapshotChangeRecordEmitter(P partition, OffsetContext offset, Object[] row, Clock clock, CommonConnectorConfig connectorConfig) {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/SnapshotChangeRecordEmitter.java:23]: @ironakj Can we use RelationalDatabaseConnectorConfig here?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlChangeRecordEmitter.java [hunk 6]
```diff
@@ -27,16 +28,14 @@ public class MySqlChangeRecordEmitter extends RelationalChangeRecordEmitter<MySq
     private final OffsetContext offset;
     private final Object[] before;
     private final Object[] after;
-    private final boolean skipMessagesWithoutChange;
 
     public MySqlChangeRecordEmitter(MySqlPartition partition, OffsetContext offset, Clock clock, Operation operation, Serializable[] before,
-                                    Serializable[] after, boolean skipMessagesWithoutChange) {
-        super(partition, offset, clock);
+                                    Serializable[] after, CommonConnectorConfig connectorConfig) {
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlChangeRecordEmitter.java:33]: Is it possible to use `MySqlConnectorConfig` here and the appropriate config types at other places too use the most specific type?

**Review by jpechane — CHANGES_REQUESTED** [review]
@ironakj Thanks a lot. I left comments related the useage of specific types. Could you please also prepare PRs for non-core connector repos?

File: debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java (hunk 4 above)
Comment (ironakj) [reply debezium-core/src/main/java/io/debezium/relational/RelationalChangeRecordEmitter.java]: Done

File: debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java (hunk 3 above)
Comment (ironakj) [reply debezium-core/src/main/java/io/debezium/pipeline/AbstractChangeRecordEmitter.java:30]: Marking this as resolved

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java (hunk 2 above)
Comment (ironakj) [reply debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java:206]: Ok

**Pull request merged**